### ISSUE TITLE
openwrt: revive/update the package?

### ISSUE THREAD

**Issue by bdklahn** [peervpn/peervpn#15]
Thanks for creating this simple to use, yet powerful tool! I like how simple it is to configure more than one init peer, to add as much discovery redundancy as you want/need.

I created Issue #3400 at https://github.com/openwrt/packages, asking whether they might update, and again support, a current version of peervpn. I was able to use an older, unsupported, legacy package (peervpn_0-023-1_brcm47xx.ipk) on their current release (15.05) on my device. But I figured an updated one might be more prudent. I proposed a potential updated Makefile for them, and they asked if I want to become the package maintainer. I'm still considering, but I wondered if @tobiasvolk , or someone else here, might want to have control of this, instead.
At least, I'd like to know maybe if/how Tobias (et. al.?) might like to be acknowledged in, maybe, the package description.

I thought it might be nice to use peervpn as a secure gateway into my firewalled/NAT'd network. This reminds me about a question (which might be addressed in a README, or text note in the example config file). How would I do that? OpenVPN, I believe, has a "gateway" option/setting, which, for example, allows one to ping from the VPN address space, into the lan. How does/might this be done with peervpn? Can I just set addresses consistent with the lan? Would I just create a software bridge interface between the lan and the tap device? It's probably simple. I just don't know as much as I could about it.

Thanks for your time!

**Comment by wankdanker** [comment]
Hey @bdklahn, I'm the one who initially submitted the package to OpenWRT. I would have continued to add new releases, but the process of getting packages accepted seemed tedious at the time (I don't think they were on GitHub then; had to email patches). Not sure if you found it, but last Makefile I could find is [here](https://git.openwrt.org/?p=14.07/packages.git;a=blob;f=net/peervpn/Makefile;h=ff930e6ff43f6e8069f0e34440303c473ca810fb;hb=HEAD).

I asked Tobias for permission to submit the package to OpenWRT and he said go for it. It didn't appear as if there was any preference for control over package maintainership on OpenWRT. Tobias may have changed his mind on this though. I personally don't have the time or interest to do it at this time.

**Comment by bdklahn** [comment]
Thanks for the response @wankdanker. Haha. I had asked some questions in that issue's comments, trying to deduce how "tedious" it might be to be a maintainer. --if I'd have to fork and use merge requests (which might be fine), or If I'd get direct push access. --if pushes would be compile tested on their build farm (or whatever), or if I'd be expected to test cross-compile locally, for all supported platforms, before they'd consider merging in an update, etc.
I had found an older Makefile, but hadn't cloned/explored all the branches/repos and found that newer version. Looks like that adds/includes the checksum and SUBMENU directives, they mentioned I'd need. Thanks!

Ok . . . well . . . I'm not positive I have the time or enough interest either, but, I might, if they have the process as streamlined/automated as it can be. Maybe @tobiasvolk can chime in, briefly. I guess it could add some visibility, and potential contributors, to peervpn, with presumeably many routers running openwrt. The "available packages" section does seem to display the "description" in the GUI. This could(?) be written to reference the efforts here.

How was the package issue support, during your tenture as maintainer? Did you get many e-mails?

**Comment by tobiasvolk** [comment]
Hi!

I believe it is always better if distro packages are maintained by the people who actually use them. So, regarding the OpenWRT package: Go ahead! :-)

Regarding the gateway setup: Yes, this is possible, and can be implemented with policy routing. This works by assigning the peervpn process a routing table that routes packets only via the physical interfaces, while all other processes get a table with (default) routes via the PeerVPN tunnel.

The implementation of this depends on your OS. In Linux, check out the "fwmark" match of the "ip rule" command and the "sockmark" option in peervpn.conf. In FreeBSD, check out "setfib(1)". I guess I should write a tutorial about this, too. :-)

- Tobias

**Comment by bdklahn** [comment]
Thanks for the response, @tobiasvolk !

Hmmm . . . using it like a gateway might be prerequisite for me to continue to "actually use" peervpn (and feel it worth being the/an OpenWRT package maintainer). I was hoping to access http servers on some Android (my old smartphones) dhcp (static assigned ip addresses) clients of that OpenWRT router (without trying to somehow run peervpn on each of them).

I kinda get what you mean with the policy routing, but would need to spend some time reading up on how to do that . . . I actually work/develop on FreeBSD (PC-BSD), but my OpenWRT router is Linux, of course

It would be ideal to have a cookbook example/tutorial, but . . .
I don't know about "should"s, for people who are already good enough to share their software with the public. And there is something to be said for tools which are maybe designed not to necessarily need that much documentation. When I read the example config file, with the comments, it became pretty clear what you could do with peervpn, and how to do it. I liked that simplicity/clarity.

But yeah . . . maybe some examples in a README.md, maybe a couple more comments in the example config file.

Thanks,
Brian

**Comment by bdklahn** [comment]
I might as well close this issue, here. Either way, it looks like this wouldn't be controlled/happening from this end.

My question about using the tunnel as/like a gateway probably was more appropriate to separate into another issue, maybe asking for some limited/concise how-to documentation, somewhere. But since I raised that here, I might as well let others know how I ended up dealing with it, in my case.
Basically, in OpenWRT, I ended up adding the peervpn-created tap0 device to the lan bridge (which already bridged the wifi interface, with the physical switch ports). I added a minimal tap0 device section to the network config file, just to get it to show up in the LuCI web interface. Then I was able to check the box to add that to the bridge (I could have also done that in /etc/config/network). Before this, though, I set the peervpn config's address space to be the same as my lan. The addresses used for each end of the tunnel are in a range NOT allowed for the dhcp server to assign to. I don't know if this bridge configuration is possible or recommended on non-Broadcom-based chips. I hope I haven't somehow inadvertently made my setup unacceptably insecure. With my current firewall setup, NAT, etc., I'm assuming the only way I'm now able to connect to my internal lan is through the tap0 peervpn tunnel. If that is the case, I then assume all this traffic is encrypted ("enabletunneling yes").
On OpenWRT, it looks like routing (including policy-based, with marked packets) is also handled in /etc/config/network. So what @tobiasvolk mentioned, should be possible. The OpenWRT documentation seems ambiguous/sparse/outdated enough to not allow deduction of exactly how to do that, without some trial and error.
